### PR TITLE
doc for external helpers

### PR DESCRIPTION
**Pull request by paulr34**
Adding readMe for instructions on how to create External Helpers

### CODE REVIEW

File: docs/external-helpers.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+## Introduction
+
+This guide explains how to add and manage externalized helper functions in the SDK, which interacts with ZAP (ZCL Advanced Platform). Initially, the SDK helpers were external to the ZAP repository, but due to maintenance and design concerns, they were moved into the ZAP repository. With the introduction of a new design pattern, SDK developers now have the flexibility to create their own externalized helpers and integrate them without requiring access to ZAP's source code.
+
+This design enables SDK developers to specify helper functions via the `gen-template.json` configuration file, which ZAP consumes during binary creation. The SDK can then call these helpers, passing in the necessary data, making them more flexible and reducing dependency on the internal ZAP codebase.
+
+## Background
+
+### Previous Approach
```
Comment (tecimovic) [inline docs/external-helpers.md:11]: You can leave this here for a while, but ultimately, nobody cares about "previous approach".

File: docs/external-helpers.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+In the earlier design, SDK helpers were external to ZAP. However, to function correctly, these helpers required access to ZAP’s source code, especially certain files within the ZAP repository. This created challenges:
+
+- When ZAP function names changed, corresponding updates had to be made across the SDK.
+- SDK developers were constrained by these internal dependencies, requiring explicit imports or "requires" of ZAP files.
+
+### New Design Pattern
+
+The new design pattern aims to free SDK developers from being tied to the internal structure of the ZAP repository, while still allowing the SDK to interact with ZAP's API. The new design works as follows:
```
Comment (tecimovic) [inline docs/external-helpers.md:20]: This should be "The design pattern", not "New design pattern".

Documentation should care about "what's there now", not "how did we get here".

File: docs/external-helpers.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+```javascript
+/**
+ * Returns an array of attributes for a given cluster.
+ * The cluster input is required to come from a script-api in this module.
+ *
+ * @param {*} context
+ * @param {*} endpoint
+ * @param {*} cluster
+ */
+function attributes(context, endpoint, cluster) {
```
Comment (tecimovic) [inline docs/external-helpers.md:39]: I don't understand this code snippet.

File: docs/external-helpers.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+### Step 2: Specify Helpers in gen-template.json
+
+The gen-template.json file allows SDK developers to specify the path to the externalized helpers they wish to use. The file should include the helper file's relative path within your project structure.
+
+For example, your gen-template.json might include:
+
+```javascript
+{
+  "helpers": [
+    "path/to/external_helpers/get_total_events_helper.js",
```
Comment (tecimovic) [inline docs/external-helpers.md:83]: You should say explicitely that this is a relative path from the directory where the gen-template.json file itself is located.

**Review by tecimovic — APPROVED** [review]
Some minor changes, but generally ok.

File: docs/external-helpers.md (hunk 1 above)
Comment (paulr34) [reply docs/external-helpers.md:11]: Yeah I agree

File: docs/external-helpers.md (hunk 2 above)
Comment (paulr34) [reply docs/external-helpers.md:20]: Fair enough, I'll remove it

File: docs/external-helpers.md (hunk 4 above)
Comment (paulr34) [reply docs/external-helpers.md:83]: Okay will do

File: docs/external-helpers.md (hunk 3 above)
Comment (paulr34) [reply docs/external-helpers.md:39]: Thanks I'll add more clarification

File: docs/external-helpers.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    "path/to/external_helpers/get_total_attributes_helper.js"
+  ]
+}
+```
+
+The paths should be relative to the location of the gen-template.json file itself. This tells ZAP where to find your helper files when generating the binary.
+
+### Step 3: Register Helpers with ZAP
```
Comment (ethanzhouyc) [inline docs/external-helpers.md:69]: Do we need to register helpers in a specific file? If so, showing the file here would help

Comment (paulr34) [reply docs/external-helpers.md:69]: No you just create a file and point to it in gem-templates.json

File: docs/external-helpers.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  )
+}
+```
+
+- The `helperRegister.registerHelpers()` function registers each helper.
+- The first argument is the **name** of the helper.
+- The second argument is the **helper function** you defined.
+- The third argument is the **context object**, which contains the necessary environment or data (such as a database connection) that may be passed to the helper.
+
+### Step 4: Use the Helpers in Your SDK Code
```
Comment (ethanzhouyc) [inline docs/external-helpers.md:98]: So after registration, the helper function will be global ready to use in any file without import?

**Review by ethanzhouyc — APPROVED** [review]

**Pull request merged**